tooling/templatize/pkg/pipeline/types.go

```diff
@@ -14,25 +16,170 @@ type Pipeline struct {
 }
 
 type ResourceGroup struct {
-	Name         string  `yaml:"name"`
-	Subscription string  `yaml:"subscription"`
-	AKSCluster   string  `yaml:"aksCluster,omitempty"`
-	Steps        []*Step `yaml:"steps"`
+	Name         string `yaml:"name"`
+	Subscription string `yaml:"subscription"`
+	AKSCluster   string `yaml:"aksCluster,omitempty"`
+	Steps        []Step `yaml:"steps"`
+}
+
+func (rg *ResourceGroup) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	rawRg := &struct {
+		Name         string    `yaml:"name"`
+		Subscription string    `yaml:"subscription"`
+		AKSCluster   string    `yaml:"aksCluster,omitempty"`
+		Steps        []rawStep `yaml:"steps"`
+	}{}
+	if err := unmarshal(&rawRg); err != nil {
+		return err
+	}
+	rg.Name = rawRg.Name
+	rg.Subscription = rawRg.Subscription
+	rg.AKSCluster = rawRg.AKSCluster
+	rg.Steps = make([]Step, len(rawRg.Steps))
+	for i, rawStep := range rawRg.Steps {
+		switch rawStep.meta.Action {
+		case "Shell":
+			rg.Steps[i] = &ShellStep{}
+		case "ARM":
+			rg.Steps[i] = &ARMStep{}
+		default:
+			return fmt.Errorf("unknown action type %s", rawStep.meta.Action)
+		}
+		err := rawStep.unmarshal(rg.Steps[i])
+		if err != nil {
+			return err
+		}
+	}
+	return nil
 }
 
 type outPutHandler func(string)
 
-type Step struct {
-	Name            string     `yaml:"name"`
-	Action          string     `yaml:"action"`
+type StepMeta struct {
+	Name      string   `yaml:"name"`
+	Action    string   `yaml:"action"`
+	DependsOn []string `yaml:"dependsOn,omitempty"`
+}
+
+func (m *StepMeta) StepName() string {
+	return m.Name
+}
+
+func (m *StepMeta) ActionType() string {
+	return m.Action
+}
+
+func (m *StepMeta) Dependencies() []string {
+	return m.DependsOn
+}
+
+type rawStep struct {
+	meta      *StepMeta
+	unmarshal func(interface{}) error
+}
+
+func (msg *rawStep) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	msg.meta = &StepMeta{}
+	if err := unmarshal(msg.meta); err != nil {
+		return err
+	}
+	msg.unmarshal = unmarshal
+	return nil
+}
+
+func (msg *rawStep) Unmarshal(v interface{}) error {
+	return msg.unmarshal(v)
+}
+
+type Step interface {
```

Comment:
No `Run()`?

Reply:
i had a `Run` in there to begin with but then reconsidered. i have high hopes that these structures will be the ones we will harmonize onto when we merge the code from ARO-HCP + ADO-Pipelines into the new repo. and for the EV2 context, a run function does not make any sense. so  i tried to keep them as data-only as possible.